File: matlab/secantmethod.m
clear all
clc

tic
% Functions
func = @(x) x^2 + 4 * cos(x);   % algorithm function
d_func = @(x) 2*x - 4 * sin(x); % derivative of function

% Secant Method Algorithm
eps = 10^(-4) ;
x_old = 1;
iter = 0;
error = 100;
while (error > 10^-5)
    
    dd_func = (d_func(x_old + eps) - d_func(x_old - eps))/(2*eps); % centered difference approximation
    x_new = x_old - d_func(x_old)/dd_func; % new x value
     
    error = abs(x_new - x_old); % tolarence 
    x_old = x_new;     % update x(i)       
    iter = iter + 1;   % update iteration number
      
end
time = toc;
% Print results
fprintf("\tSecant Method\n")
fprintf("Iteration : %d\n",iter);
fprintf("Computation time : %f\n",time);
fprintf("Minimum point of function : x = %f, f(x) = %f\n",x_new,func(x_new));

File: c++/secant_method.cpp
#include <bits/stdc++.h>

#define MAX 99999

double func(double x){
    // Function to minimize, this can be changed to any function.
    // Parameters:
    //      :x: variable of function
    return x * x + 4 * cos(x);   
}


double dfunc(double x){
    // Derivative of minimized function
    return 2 * x - 4 * sin(x);
}


double secant(double xPrev){

    std::clock_t start;
    start = std::clock();

    int iter = 0;
    double eps = 0.0001; // epsilon value for derivation
    double xNew, error = MAX;

    while (error > exp10(-5)){

        double ddfunc = (dfunc(xPrev + eps) - dfunc(xPrev - eps)) / (2*eps); // centered difference approximation
        xNew = xPrev - dfunc(xPrev) / ddfunc;         
        error = fabs(xNew - xPrev);
        xPrev = xNew;
        iter++;
    }

    std::cout << "Secant Method" << std::endl;
    std::cout << "Iteration : " << iter << std::endl;
    std::cout << "Computation Time : "  << std::clock() - start << std::endl;
    std::cout << "Minimum point of function : " << "x = " << xNew  << ", f(x) = " << func(xNew) << std::endl; 

    return xPrev;
}


int main(){
    secant(1.00);



    return 0;
}